src/Bicep.Core/Semantics/Namespaces/AzNamespaceType.cs

```diff
@@ -348,6 +337,17 @@ private static IEnumerable<FunctionOverload> GetAzOverloads(ResourceScope resour
             }
             else
             {
+                foreach (var (functionOverload, allowedScopes) in GetScopeFunctions())
```

Comment:
I'm not suggesting you make this change now (more interested in just getting your thoughts on it), but in the long run, I've been wondering if it would be less error-prone (and possibly better user experience) to have a "usage" property on the function overload - e.g.:

```csharp
new FunctionOverloadBuilder("resourceGroup")
    .WithReturnResultBuilder(GetResourceGroupReturnResult, new ResourceGroupScopeType(Enumerable.Empty<FunctionArgumentSyntax>(), Enumerable.Empty<TypeProperty>()))
    .WithGenericDescription(resourceGroupGenericDescription)
    .WithDescription("Returns the current resource group scope.")
    .WithPermittedUsage(FunctionUsage.BicepFileOnly)
// or     .WithPermittedUsage(Usage.BicepParamsFileOnly)
    .Build(),
```

AFAIK we already have a similar capability to restrict the usage of `getSecret()` which we might be able to generalize more.

It also feels like it would generally be useful to shift in the direction of separating out "static" namespace functionality, rather than throwing away and rebuilding the namespace on every keypress (when profiling the LS, I this appeared to be one of the most expensive operations we do).

Reply:
That approach makes sense to me, but we should be able to hold on to copies of static namespace types across compilations even with the current approach. Feature-flagged and file format-specific functionality will vary by path, but we would only need to rebuild namespace types in the LS if a `bicepconfig.json` file were added or modified.

If we wanted to just hold onto a singleton, I think we would need to add a `.WithRequiredFeatures(Func<IFeatureProvider, bool> predicate)` method in addition to `.WithPermittedUsage(FunctionUsage usage)`.